libghdl_pkg.py.in

```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env python
+
+from os import mkdir
+from os.path import join
+from shutil import copy, copytree
+
+
+libghdl_filename = 'libghdl-' + "@libghdl_version@" + '@SOEXT@'
+dstdir = 'build-libghdl'
+
+# Copy python sources
+copytree(
+    join('@srcdir@/src/vhdl/python/libghdl'),
+    join(dstdir),
+)
+
+# Create config file
+with open(join(dstdir, 'config.py'), 'w') as f:
+    f.write('libghdl_filename="{}"\n'.format(libghdl_filename))
+
+# Copy shared library and VHDL libraries
+copy(libghdl_filename, dstdir)
+copy('setup.py', dstdir)
+
+# Copy VHDL libraries
+def ignore_function(_, lst):
+    return [item for item in lst if item.endswith('.o')]
+copytree(
+    join('lib', 'ghdl'),
+    join(dstdir, 'ghdl'),
+    ignore=ignore_function
+)
```

Comment:
So, no need to copy the VHDL libraries, they are already installed.

Reply:
I don't know if it is worth. Correct me if I am wrong, but this PR creates two completely independent tarballs. A user can decide to install GHDL only, libghdl-py only, or both of them. For libghdl to use GHDL libraries, it should not only check that the libraries exist, but also that the version of GHDL that generated them is compatible, isn't it? If so, I think that this can lead to bad UX if users install GHDL through the package manager of their distro but then gets libghdl independently.

Furthermore, I checked that the size of the 'common resources' is around 5MB. Is it worth the effort?

Related to this issue, there is something I am not sure about. When GHDL is configured with `--enable-python`, is there any internal difference in the generated GHDL binary? I.e., does GHDL include the features of libghdl so it would be possible for python scripts to communicate with the GHDL instead of using the shared library? Or does the shared library include features which GHDL does not and also lack others?

Reply:
There is no way to add libraries with libghdl, so if you want to add an external library (like xilinx unisim) you're stuck.  Hence the need to use the same libraries as ghdl.  libghdl only needs to access to the sources, not to the binaries.

`--enable-python` should have no effect on the ghdl binary.
But libghdl has some extra features (used by the language server).

Reply:
> libghdl only needs to access to the sources, not to the binaries.

With the current procedure not only sources are being copied, `*.cf` objects are being copied too, because only `*.o` extensions are ignored. Other files are copied too: `grt.lst`, `grt.ver`, `libgrt.a`. Are any of these required in the libghdl python package?

> There is no way to add libraries with libghdl, so if you want to add an external library (like xilinx unisim) you're stuck. Hence the need to use the same libraries as ghdl. 

Then, yes, it makes sense for libghdl to use the installed libraries if available. However, I would not convert the strong suggestion of having GHDL installed into a dependency. I.e., libghdl should search locally, then an available GHDL and (optionally) some custom paths.

Nonetheless, if libghdl depends on GHDL, what's the point of making the generation of `libghdl*.so` optional? Instead of distributing a bunch of python scripts with a single shared library, I would make `libghdl*.so` part of the default GHDL installation, and let python scripts find it, the same way they are going to find the sources of VHDL libraries.

If using `--enable-python` has no effect on the binary and has no extra build dependencies:

- Remove option `--enable-python` from the codebase.
- Always build `libghdl*.so`
- Move the 'minimal' `setup.py` proposed in this PR to `src\vhdl\python`.
- Move `src\vhdl\python` to a separate directory or just tell users to treat it as if it was an independent python project.

> But libghdl has some extra features (used by the language server).

But it does also lack features from the binary, i.e., I cannot use libghdl to analyze and elaborate designs, can I?

Reply:
Only .cf and vhdl source files are required.  grt is not.

Looking on multiple paths would be a solution for the problem.  But why you don't like the dependency on the ghdl binary ?

Although we could make libghdl*.so part of the default installation, it is not obvious how it will be found.  The easiest solution would be keep the config.py file, but then it wouldn't be fully independent.

To sum up:
1) we need libghdl to be able to use the libraries from ghdl.
2) we could always build libghdl (or maybe add --disable-libghdl)
3) it would be nice to streamline the libghdl python part, and ideally it would be nice if I could use `pip install -e` on it (but I don't know how to do that).

Reply:
You could use libghdl to analyze, but right now you cannot elaborate designs with it.

Reply:
For this PR: just move libghdl_pkg.py to the makefile and I will merge it.

Reply:
> Only .cf and vhdl source files are required. grt is not.

OK. I will clean it up after I get a clear picture.

> Looking on multiple paths would be a solution for the problem.

I think that this is an independent but related issue. I mean, on some contexts the user might have a repository/directory of sources/libraries that are not installed with GHDL, but which are used in different projects.

> But why you don't like the dependency on the ghdl binary ?

It is not that I don't like it. I'm perfectly ok with it, if that's the way it needs to be. I'm just struggling to understand where are the tight and loose links between the binary and the shared library. In the end, it is also possible to load the GHDL binary in Python using the same library/function: https://github.com/ghdl/ghdl/blob/master/src/vhdl/python/libghdl/__init__.py#L7.

So, if libghdl needs GHDL to be installed *1 on the system, is it really worth building `libghdl*.so` at all? Why not add the additional features to the binary itself and allow to load it dynamically? As a result, libghdl would be a Python-only package, independent from the target platform/architecture.

*1 This is because it is not enough with VHDL sources, `*.cf` files are also needed and these are only generated through the binary, isn't it?

Note that I am arguing in favor of not distributing the `*.so` library as a separate artifact because it is platform-dependent, as the binary. Therefore, should we want to publish it with regular releases, we would duplicate the number of artifacts. Conversely, if we distribute the binary and the shared library in the same package (either in separate files, or with the features merged in a single object), we can provide libghdl with regular releases by adding just a tarball with python sources. It might need a config file, but it will work on any platform.

> Although we could make libghdl*.so part of the default installation, it is not obvious how it will be found. The easiest solution would be keep the config.py file, but then it wouldn't be fully independent.

If libghdl needs to know the location of GHDL libraries, it is possible to get the location of the GHDL binary too, and of any other resource in the package, isn't it? I need to have a better look at the codebase affected by this...

From another point of view, VUnit does actually search and locate ghdl if it is available in the PATH, or set through an envvar. We can follow the same procedure here.

Nonetheless, I think that the important point is to reduce the complexity of the build/packaging/installation procedure. I think that keeping packages separated  by language/framework would help:

- GHDL: Ada, VHDL and C
- libghdl: Python
- ghdl-language-server/ghdl-ls: Python
- vscode-client: TypeScript

I don't suggest to create four repos; two is ok, because GHDL/libghdl and ghdl-ls/vscode-client are very tightly coupled. Moreover, all of them could be located in the main repo. That's not a problem I think.

Note that, in any case, there would be two 'compiled packages' (GHDL and vscode extension) and two 'sources-only' packages (libghdl and ghdl-ls).

> we could always build libghdl (or maybe add --disable-libghdl)
>
> it would be nice to streamline the libghdl python part, and ideally it would be nice if I could use pip install -e on it (but I don't know how to do that).

- I will close #839, because it is part of this PR anyway.
- I will keep this PR on hold until I do some tests with the packaging procedure (regarding `--disable-ghdl`).
- If the scheme I am thinking about works, we won't need `libghdl_pkg.py` at all.

Reply:
> For this PR: just move libghdl_pkg.py to the makefile and I will merge it.

I don't get very well with makefiles. I will do the tests with making libghdl use GHDL libraries and building libghdl by default. If I get stuck with any of them, I will modify the makefile to have this merged at least.

Reply:
I fear that loading ghdl from python is an hack that doesn't work on Windows.  We could add the missing features in ghdl, but there would still be the issue with Windows.

I am OK with building libghdl.so with ghdl, and keeping libghdl.py as pure as possible.  Note that they are tightly coupled, so they have to be kept in sync.

The `.cf` files are needed to know where are the vhdl source files.

I also agree that libghdl.py could automatically find libghdl.so.  Just a matter of setting the location in config.py

So I think we agree!

For copying the library files for libghdl.py, I fear we missed something.  Right now they are installed with setup.py, so along side libghdl.py.  If the copy is made by a different tool (either libghdl_pkg.py or makefile) then the destination is not known anymore.

OTOH, if libghdl.so is part of ghdl, no need to have two copies of the libraries.  Sorry to raise this point again, but this is not fully clear.

Reply:
> I fear that loading ghdl from python is an hack that doesn't work on Windows. We could add the missing features in ghdl, but there would still be the issue with Windows.

You are correct. On windows, EXEs and DLLs are not ELFs. Currently, is libghdl built as a DLL? If so, for the future it might be possible to use a single binary on Linux and EXE + DLL on Windows. Nevertheless, we will forget about this for now.

> I am OK with building libghdl.so with ghdl, and keeping libghdl.py as pure as possible. Note that they are tightly coupled, so they have to be kept in sync.

Yes. The tests I want to make are mostly related to how to ensure that we don't get things mixed up. Have you considered releasing the Python-only packages 'libghdl' and 'ghdl-ls' through PyPi (as e.g. VUnit)? I'm not suggesting that we should do it, just asking about it.

> The `.cf` files are needed to know where are the vhdl source files.

But, can these be generate by libghdl (by analyzing the sources) or they can only be generated with the GHDL binary?

> I also agree that libghdl.py could automatically find libghdl.so. Just a matter of setting the location in config.py
> 
> So I think we agree!

We do! We always do; sometimes it just takes some more time ;)

> For copying the library files for libghdl.py, I fear we missed something. Right now they are installed with setup.py, so along side libghdl.py. If the copy is made by a different tool (either libghdl_pkg.py or makefile) then the destination is not known anymore.
> 
> OTOH, if libghdl.so is part of ghdl, no need to have two copies of the libraries. Sorry to raise this point again, but this is not fully clear.

The installation, if done, is still done with either pip or by executing setup.py. The difference is that now the setup.py function takes/copies sources from very different paths. The proposal in this PR is to gather all those resources in a directory manually (or through the makefile), and then copy a corresponding setup.py file into that directory.

So, indtead of letting setup.py copy from `src/vhdl/python/libghdl` and `lib/ghdl`. We copy them to `build/libghdl` and `build/ghdl`, respectively, and then let setup.py take them from there.

Nonetheless, I'm not sure about understanding what you are trying to explain here... If you mean that subdir `ghdl` might need to be added to option `package_data=` in the `setup` function, then yes, you might be correct. I had to had the `*.so` file explicitly, because it copied Python sources only.

However, after moving libghdl.so to be part of GHDL, the will be no VHDL sources at all alongside libghdl.py. Precisely, the first thing that will do libghdl.py is to look for GHDL in order to find the shared library first and the .cf files and VHDL sources when needed.

Reply:
So, yes libghdl.so is built as a DLL.  So it should work on windows (I have never tested it).

It would be nice to be able to distribute libghdl.py through Pypi; but as it is tightly coupled with ghdl this could be done only for versions.  I don't have a strong opinion.  Let's wait.

Libghdl has all the code to generate the `.cf` files, but there is currently no tool for that.

To clarify the installation:
1) libghdl.so is now part of ghdl.
2) there is no vhdl sources at all alongside libghdl.py
3) libghdl.py is 'almost' pure and is installed by setup.py or pip.